staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go

```diff
@@ -87,43 +87,56 @@ func (a customResourceStrategy) PrepareForCreate(ctx context.Context, obj runtim
-	if !utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceSubresources) || a.status == nil {
-		return
-	}
-
 	newCustomResourceObject := obj.(*unstructured.Unstructured)
 	oldCustomResourceObject := old.(*unstructured.Unstructured)
 
 	newCustomResource := newCustomResourceObject.UnstructuredContent()
 	oldCustomResource := oldCustomResourceObject.UnstructuredContent()
 
-	// update is not allowed to set status
-	_, ok1 := newCustomResource["status"]
-	_, ok2 := oldCustomResource["status"]
-	switch {
-	case ok2:
-		newCustomResource["status"] = oldCustomResource["status"]
-	case ok1:
-		delete(newCustomResource, "status")
+	var incrementGeneration bool
+	if utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceSubresources) && a.status != nil {
+		// If the /status subresource endpoint is installed, it
+		// indicates that the user wants to participate the spec/status
+		// convention, then only changes to `.spec` increments the
+		// `.metadata.generation`
+		newSpec, ok1 := newCustomResource["spec"]
+		oldSpec, ok2 := oldCustomResource["spec"]
+
+		// spec is changed, created or deleted
+		if (ok1 && ok2 && !apiequality.Semantic.DeepEqual(oldSpec, newSpec)) || (ok1 && !ok2) || (!ok1 && ok2) {
+			incrementGeneration = true
+		}
+	} else {
+		// except for the changes to `metadata`, any other changes
+		// cause the generation to increment.
+		newObjCopy := newCustomResourceObject.DeepCopy()
+		oldObjCopy := oldCustomResourceObject.DeepCopy()
+		newCopyContent := newObjCopy.UnstructuredContent()
+		oldCopyContent := oldObjCopy.UnstructuredContent()
+		delete(newCopyContent, "metadata")
+		delete(oldCopyContent, "metadata")
+		if !apiequality.Semantic.DeepEqual(newCopyContent, oldCopyContent) {
+			incrementGeneration = true
+		}
 	}
 
-	// Any changes to the spec increment the generation number, any changes to the
-	// status should reflect the generation number of the corresponding object. We push
-	// the burden of managing the status onto the clients because we can't (in general)
-	// know here what version of spec the writer of the status has seen. It may seem like
-	// we can at first -- since obj contains spec -- but in the future we will probably make
-	// status its own object, and even if we don't, writes may be the result of a
-	// read-update-write loop, so the contents of spec may not actually be the spec that
-	// the CustomResource has *seen*.
-	newSpec, ok1 := newCustomResource["spec"]
-	oldSpec, ok2 := oldCustomResource["spec"]
-
-	// spec is changed, created or deleted
-	if (ok1 && ok2 && !apiequality.Semantic.DeepEqual(oldSpec, newSpec)) || (ok1 && !ok2) || (!ok1 && ok2) {
+	if incrementGeneration {
 		oldAccessor, _ := meta.Accessor(oldCustomResourceObject)
 		newAccessor, _ := meta.Accessor(newCustomResourceObject)
 		newAccessor.SetGeneration(oldAccessor.GetGeneration() + 1)
 	}
+
+	// If the /status subresource endpoint is installed, update is not allowed to set status.
```

Comment:
It seems like if you do this before the if/else branch that sets `incrementGeneration`, you can delete everything in the `if` branch and just do the stuff in `else` every time, assuming that you accept the suggestion from @smarterclayton on what Generation means when the resource has opted into spec/status convention.